### PR TITLE
MM-45494: Edit message history

### PR DESCRIPTION
**Pull request by AshishDhama**
#### Summary
This PR introduces a new feature for users to 
-	see the edit history of their edited messages
-	restore an old version

#### Ticket Link
Fixes https://github.com/mattermost/mattermost-server/issues/20864
JIRA: https://mattermost.atlassian.net/browse/MM-45494

#### Related Pull Requests

- - Has webapp changes (https://github.com/mattermost/mattermost-webapp/pull/11555)
#### Screenshots

https://user-images.githubusercontent.com/37421564/189528414-5afc2363-a227-4cb9-9221-fc19a0ac6631.mov

#### Release Note
```release-note
Give ability to users to see the history of edited messages and restore an old version with current version.
```

### CODE REVIEW

**Comment by AshishDhama** [comment]
@sinansonmez could you please take a look and confirm everything is fine? Have merged the latest master.

File: store/sqlstore/post_store.go [hunk 1]
```diff
@@ -2339,6 +2339,31 @@ func (s *SqlPostStore) GetPostsByIds(postIds []string) ([]*model.Post, error) {
...
+func (s *SqlPostStore) GetEditHistoryForPost(postId string) ([]*model.Post, error) {
+	builder := s.getQueryBuilder().
+		Select("*").
+		From("Posts").
+		Where(sq.Eq{"Posts.OriginalId": postId}).
+		OrderBy("Posts.EditAt DESC")
+
+	queryString, args, err := builder.ToSql()
+	if err != nil {
+		return nil, errors.Wrap(err, "postsForExport_toSql")
```
Comment (vish9812) [inline store/sqlstore/post_store.go:2351]: I think this(`postsForExport_toSql`) got wrongly copied. Maybe you can add something like this `failed to find post history` instead?

File: store/sqlstore/post_store.go [hunk 2]
```diff
@@ -2339,6 +2339,31 @@ func (s *SqlPostStore) GetPostsByIds(postIds []string) ([]*model.Post, error) {
...
+	}
+
+	posts := []*model.Post{}
+	err = s.GetReplicaX().Select(&posts, queryString, args...)
+	if err != nil {
+		return nil, errors.Wrapf(err, "error getting posts edit history with postId=%s", postId)
```
Comment (vish9812) [inline store/sqlstore/post_store.go:2357]: A check for notFound error will be needed here.
```go
if err == sql.ErrNoRows {
  return nil, store.NewErrNotFound("Post", postId)
}
```

**Review by vish9812 — COMMENTED** [review]
It looks good to go, other than a couple of changes needed.
Also, a test case for store function will be needed.

File: api4/post_test.go [hunk 3]
```diff
@@ -3076,6 +3076,73 @@ func TestGetPostsByIds(t *testing.T) {
...
+	require.NoError(t, err2)
+	CheckOKStatus(t, response2)
+
+	t.Run("update history correctly", func(t *testing.T) {
+		history, response3, err3 := client.GetEditHistoryForPost(rpost.Id)
+		require.NoError(t, err3)
+		CheckOKStatus(t, response3)
+
+		require.Len(t, history, 2)
+		require.Equal(t, history[0].Message, "new message edited")
```
[the remark this answers is not in the document]
Comment (sinansonmez) [reply api4/post_test.go:3127]: Good catch, thanks

**Comment by sinansonmez** [comment]
@mylonsuren tests are updated as suggested

File: api4/post.go [hunk 4]
```diff
@@ -494,6 +495,39 @@ func getPostsByIds(c *Context, w http.ResponseWriter, r *http.Request) {
...
+
+	if c.AppContext.Session().UserId != originalPost.UserId {
+		c.SetPermissionError(model.PermissionEditPost)
+		return
+	}
+
+	if !c.App.SessionHasPermissionToChannelByPost(*c.AppContext.Session(), c.Params.PostId, model.PermissionEditPost) {
+		c.SetPermissionError(model.PermissionEditPost)
+		return
+	}
```
Comment (mylonsuren) [inline api4/post.go:518]: I think the ordering could change here, might lead to misleading errors when debugging in the future. We can also  pass through the original error from `GetSinglePost`.

```suggestion
	if !c.App.SessionHasPermissionToChannelByPost(*c.AppContext.Session(), c.Params.PostId, model.PermissionEditPost) {
		c.SetPermissionError(model.PermissionEditPost)
		return
	}
	
	originalPost, err := c.App.GetSinglePost(c.Params.PostId, false)
	if err != nil {
		c.Err = err
		return
	}

	if c.AppContext.Session().UserId != originalPost.UserId {
		c.SetPermissionError(model.PermissionEditPost)
		return
	}
```

**Review by mylonsuren — COMMENTED** [review]
Looks good @sinansonmez, awesome work!

**Comment by sinansonmez** [comment]
kindly pinging for review @vish9812 @mylonsuren

**Review by vish9812 — APPROVED** [review]
Great work @sinansonmez 🎉

**Comment by sinansonmez** [comment]
Hi @jgilliam17 [this PR](https://github.com/mattermost/mattermost-webapp/pull/11069) is the same PR. It is just replicated by Ashish in order to create test server. Therefore you can test the PR using test server here: https://github.com/mattermost/mattermost-server/pull/21601#issuecomment-1347905539

**Comment by jgilliam17** [comment]
@sinansonmez or @AshishDhama I tried updating branch, not sure if it was successful due to [enterprise failure](https://github.com/mattermost/mattermost-server/pull/21601#issuecomment-1373887737) 
Seeing one issue on the test server
When edit history rhs is open, opening Sys console and returning to channel is causing a white screen
Screen Shot 2023-01-06 at 2 06 27 PM

**Comment by jgilliam17** [comment]
Few more issues I missed on the first pass:
1. When we are restoring long edited post, do we want to scroll text within the restore confirmation modal or leave current functionality (seen on the test server) where the whole modal expands to display long post causing modal to scroll?

https://user-images.githubusercontent.com/52937121/211087750-75c76be6-6ab4-43d6-b2ba-873b217751bc.mov

2. Channel name on the Edit History RHS updates when switching channel. We should keep the correct channel name for the post history user is viewing or close edit history RHS on channel switch 
Steps: 
Open post history rhs for post on Town Square, switch channel to Off Topic
Observed: post history channel label switches to Off Topic

@matthewbirtch Do we want to close Edit History RHS when changing the channel

**Comment by matthewbirtch** [comment]
> 1. When we are restoring long edited post, do we want to scroll text within the restore confirmation modal or leave current functionality (seen on the test server) where the whole modal expands to display long post causing modal to scroll?

Good catch @jgilliam17. We should look at what we do in the `Forward message` modal and apply the same logic for long posts. I believe we limit long posts in this modal to a certain number of lines and have a 'Show more' button to expand it.

> @matthewbirtch Do we want to close Edit History RHS when changing the channel

Another great catch @jgilliam17. I think it's reasonable to close the edit history RHS if you navigate away from the channel you're editing a post within.

**Comment by sinansonmez** [comment]
> @sinansonmez or @AshishDhama I tried updating branch, not sure if it was successful due to [enterprise failure](https://github.com/mattermost/mattermost-server/pull/21601#issuecomment-1373887737)
> Seeing one issue on the test server
> When edit history rhs is open, opening Sys console and returning to channel is causing a white screen
> Screen Shot 2023-01-06 at 2 06 27 PM

@jgilliam17 I couldnt reproduce this issue. I see `websocket closed` message just before the error. Maybe it was something related to test server. 

@AshishDhama I also updated branches (both for web and server). Could you please pull the changes from both PRs? @jgilliam17 after that you can check again.

**Comment by sinansonmez** [comment]
Great catches @jgilliam17 

> > 1. When we are restoring long edited post, do we want to scroll text within the restore confirmation modal or leave current functionality (seen on the test server) where the whole modal expands to display long post causing modal to scroll?
> 
> Good catch @jgilliam17. We should look at what we do in the `Forward message` modal and apply the same logic for long posts. I believe we limit long posts in this modal to a certain number of lines and have a 'Show more' button to expand it.

I used the same component from `Forward message` so behaviour should be the same. Please have a look @matthewbirtch  after Ashish pull the changes from my PR

> > @matthewbirtch Do we want to close Edit History RHS when changing the channel
> 
> Another great catch @jgilliam17. I think it's reasonable to close the edit history RHS if you navigate away from the channel you're editing a post within.

I also checked other places where this might be an issue. Therefore I did the changes to close the RHS in the following places. Please let me know if I am missing any potential area.

- When changing the channel from the sidebar
- When changing the channel from quick channel switcher (i.e. cmd+K)
- When you click `Join` or `Create` buttons in `More Channels` modal
- When you click `Create New Channel` or `Open a Direct Message` buttons in the plus button on the sidebar (i.e. the button in the screenshot below)

Screen Shot 2023-01-08 at 20 35 42

**Comment by matthewbirtch** [comment]
@sinansonmez thanks for catching the bug with the 'Show more' button in the forward message modal. I'll file a bug for that.

For the "restore version" modal, rather than expanding the height of the entire modal when we show more, is it possible to make it so that we increase the height of the modal to a max-height and then only the body area of the modal scrolls after that? Right now the modal gets cut off on the bottom. Maybe @AshishDhama or @michelengelen  can help with that.

image

**Comment by sinansonmez** [comment]
@matthewbirtch sure I can give it a try. What should be value for `max-height`?

**Comment by matthewbirtch** [comment]
> @matthewbirtch sure I can give it a try. What should be value for `max-height`?

I would suggest it should be the same as whatever is used in the forward message modal, but that may be broken right now. @michelengelen do you know if we added a max-height to the modal body for forward messages feature? I am seeing this calculation right now on the `modal-body` in the inspector

```
max-height: calc(100vh - 128px - 61px - 88px);
```

**Comment by michelengelen** [comment]
> > @matthewbirtch sure I can give it a try. What should be value for `max-height`?
> 
> I would suggest it should be the same as whatever is used in the forward message modal, but that may be broken right now. @michelengelen can you confirm this is the max-height of the modal body for forward messages feature? I am seeing this calculation right now on the `modal-body` in the inspector
> 
> ```
> max-height: calc(100vh - 128px - 61px - 88px);
> ```

@matthewbirtch yes, that is correct. I took this calculation from your design spec on the forwarding feature

**Comment by sinansonmez** [comment]
`max-height` is added as suggested. @AshishDhama could you please pull recent changes from webapp PR and re-apply test-server label in your server PR. 

@matthewbirtch after changes are reflected you can give a look and @jgilliam17 can test the changes

**Comment by jgilliam17** [comment]
Thanks @AshishDhama and @sinansonmez 
Looks great, just few issues remain
1. With CRT on, unable to revert to original version on thread reply (non root post) 
Note: actually, repros without CRT too

https://user-images.githubusercontent.com/52937121/214403908-d7cdf685-4935-4237-9353-63b525e2fb3c.mov



2. WIth/without CRT: When the same user is creating root post and replies, restoring version on edited reply - applies changes to root post not the reply

https://user-images.githubusercontent.com/52937121/214410823-b139d1c8-5a3a-4758-a8b6-a5f83777eb4d.mov


3. Error when opening edit history from global threads or Insights 
Screen Shot 2023-01-24 at 3 33 58 PM
4. Similar console error when switching to Insights or Threads while Edit History is open, and then opening any channel (doesn't have to be the one that had edit history previously open)

**Comment by jgilliam17** [comment]
@sinansonmez Would you be open to creating a Cypress test for edit message history?

**Comment by sinansonmez** [comment]
@jgilliam17 thanks great catches. I will look into them and also write cypress test

**Comment by jgilliam17** [comment]
@AshishDhama or @sinansonmez Can you please fix failed checks so I can trigger e2es? Thanks

**Comment by jgilliam17** [comment]
@sinansonmez Saturn brought up a good point while reviewing test case I added, about exercising permissions
Can you please add a step to the same test to check for - edit history should not available when user lacks `edit own posts` permissions

**Comment by sinansonmez** [comment]
@jgilliam17 great point. Thanks to Saturn, I forgot to include a check of `edit own post` permission in the webapp (check in server already exists). Both check and new test is added. 

After Ashish pulls recent changes from my PR, you can run e2e tests

**Comment by jgilliam17** [comment]
Thanks for adding the check for permissions 
I'll wait for Dhama to update the server for re-test. Currently there is an error that may be fixed by the changes you made on the webapp PR. Non-admin user that lacks permission to edit posts will see blank rhs when clicking on `edited` link
Additionally, re: e2e test step3
Currently it is expected that our Sysadmins can edit own and other members post, regardless of the permission set in system console.
Maybe add a step, after sysadmin changes permissions, login as non-admin to verify edit history is not available due to lack of permissions to edit own posts for non-admin users and guests. Not sure if we have to reset permissions after test is completed, checking with @furqanmlk 
Screen Shot 2023-02-01 at 4 51 29 PM

**Comment by sinansonmez** [comment]
@jgilliam17 thanks for checking

> Currently there is an error that may be fixed by the changes you made on the webapp PR. Non-admin user that lacks permission to edit posts will see blank rhs when clicking on edited link

Exactly my recent changes will fix this issue and RHS wont open at all if user dont have the permission to edit own post

> Maybe add a step, after sysadmin changes permissions, login as non-admin to verify edit history is not available due to lack of permissions to edit own posts for non-admin users and guests.

This exactly what I am testing in step 3. I am changing the config as admin. Then I am checking if non-admin user can see edit history. 
If I understood correctly, `cy.apiAdminLogin()`just log you as admin in the background but doesnt log out the current non-admin user so I dont need to login again as non-admin user. At least, this is what I saw when I was running the test in my local.

**Review by jgilliam17 — APPROVED** [review]
Thanks @sinansonmez 
LGTM
note: I created a test case, just missing the test case id. I can fill that in afterwards if one doesn't become available before merge.

**Comment by AshishDhama** [comment]
@vish9812 @mylonsuren @jgilliam17 If this is ready to be merged please do it.

**Comment by AshishDhama** [comment]
closing in favour of original pr

**Pull request closed without merging**